### PR TITLE
Fix shared sibling entry bundles

### PR DESCRIPTION
**Pull request by devongovett**
After #4421 I began experiencing #4231 again. This was due to all sibling bundles of entries being added to all bundle groups that share this asset. In reality, we only should be adding the siblings that are dependencies of the entry bundle. Bundle groups only capture what is loaded together, not the direction of the relationship.

For example, a JS bundle might reference on a sibling CSS bundle, but not the other way around. If the CSS bundle were extracted as an entry, the JS sibling shouldn't be added to the bundle groups with it.

To fix this, I've introduced a new `references` edge between bundles. When the asset type changes and we create a new bundle, a reference is created between the bundles, and we can retrieve a list of referenced bundles later on in order to add them to the relevant bundle groups.

Thanks to @wbinnssmith for the idea! 😍

### CODE REVIEW

File: packages/core/core/src/BundleGraph.js [hunk 1]
```diff
@@ -233,6 +233,10 @@ export default class BundleGraph {
     }
   }
 
+  createBundleReference(from: Bundle, to: Bundle): void {
+    this._graph.addEdge(from.id, to.id, 'references');
```
Comment (wbinnssmith) [inline packages/core/core/src/BundleGraph.js:237]: @padmaia this probably conflicts with the current use of reference edges between bundles. Where is that information used? We should probably have different edge names as these seem orthogonal.

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 7.02s | +18.00ms |
| Cached | 2.48s | +65.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 365.00ms | +111.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 364.00ms | +110.00ms ⚠️ |
| dist/legacy/styles.27b3555b.css | 29.00b | +0.00b | 1.45s | -372.00ms 🚀 |
| dist/modern/styles.27b3555b.css | 29.00b | +0.00b | 1.72s | +250.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 222.00ms | -76.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 221.00ms | -77.00ms 🚀 |
| dist/legacy/kitchen-sink.32bb4226.js | 1.13kb | +0.00b | 382.00ms | -87.00ms 🚀 |
| dist/modern/kitchen-sink.280a0cce.js | 1.13kb | +0.00b | 381.00ms | -88.00ms 🚀 |
| dist/legacy/index.html | 709.00b | +0.00b | 384.00ms | -97.00ms 🚀 |
| dist/modern/index.html | 709.00b | +0.00b | 381.00ms | -88.00ms 🚀 |
| dist/legacy/styles.27b3555b.css | 29.00b | +0.00b | 382.00ms | -88.00ms 🚀 |
| dist/modern/styles.27b3555b.css | 29.00b | +0.00b | 381.00ms | -88.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 10.73s | +30.00ms |
| Cached | 2.46s | -0.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/main/index.js | 66.80kb | +0.00b | 359.00ms | +23.00ms ⚠️ |
| dist/module/index.js | 37.45kb | +0.00b | 357.00ms | +30.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.35m | -4.35s |
| Cached | 5.10s | -45.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/component.6c698016.js | 48.66kb | +0.00b | 12.01s | +900.00ms ⚠️ |
| dist/Modal.f79bc3a9.js | 42.03kb | +0.00b | 8.60s | -565.00ms 🚀 |
| dist/component.2d909511.js | 40.07kb | +0.00b | 11.37s | -930.00ms 🚀 |
| dist/png-chunks-extract.883f0434.js | 25.63kb | +0.00b | 8.60s | -3.77s 🚀 |
| dist/DatePicker.00ddd357.js | 21.63kb | +0.00b | 27.42s | +1.93s ⚠️ |
| dist/js.786dad09.js | 16.49kb | +0.00b | 8.60s | +1.76s ⚠️ |
| dist/16.d3236fb9.js | 10.22kb | +0.00b | 10.82s | +3.99s ⚠️ |
| dist/16.c42426b9.js | 10.09kb | +0.00b | 8.60s | +1.76s ⚠️ |
| dist/16.bc89f6ee.js | 10.02kb | +0.00b | 8.60s | +1.76s ⚠️ |
| dist/16.98f3baa0.js | 9.98kb | +0.00b | 10.82s | -1.39s 🚀 |
| dist/date.1c037d1c.js | 9.04kb | +0.00b | 1.18m | +55.70s ⚠️ |
| dist/images.92799093.js | 8.98kb | +0.00b | 16.28s | +963.00ms ⚠️ |
| dist/feedback.bffa274a.js | 8.95kb | +0.00b | 27.41s | +2.00s ⚠️ |
| dist/fallback.aab3a799.js | 8.90kb | +0.00b | 19.77s | +1.32s ⚠️ |
| dist/status.97878d50.js | 8.77kb | +0.00b | 19.77s | +1.32s ⚠️ |
| dist/code.6b50d9e2.js | 8.69kb | +0.00b | 14.55s | -828.00ms 🚀 |
| dist/heading6.6daffbba.js | 8.61kb | +0.00b | 27.41s | +2.00s ⚠️ |
| dist/heading3.bcae1464.js | 8.60kb | +0.00b | 21.49s | +1.43s ⚠️ |
| dist/emoji.8c2bc3de.js | 8.54kb | +0.00b | 1.94m | +1.69m ⚠️ |
| dist/heading5.6844d092.js | 8.48kb | +0.00b | 21.49s | +1.43s ⚠️ |
| dist/expand.1a4c120f.js | 8.46kb | +0.00b | 27.41s | +2.00s ⚠️ |
| dist/heading2.a19338b9.js | 8.42kb | +0.00b | 21.49s | +1.43s ⚠️ |
| dist/mention.a261500a.js | 8.38kb | +0.00b | 18.01s | -1.72m 🚀 |
| dist/heading4.02eb6187.js | 8.37kb | +0.00b | 21.49s | +1.43s ⚠️ |
| dist/divider.acfaa3b9.js | 8.33kb | +0.00b | 1.18m | +55.68s ⚠️ |
| dist/quote.55197fd6.js | 8.33kb | +0.00b | 19.77s | +1.32s ⚠️ |
| dist/action.536bde1c.js | 8.31kb | +0.00b | 14.55s | -776.00ms 🚀 |
| dist/decision.5cdc67b3.js | 8.29kb | +0.00b | 1.18m | +55.68s ⚠️ |
| dist/panel-warning.cb245541.js | 8.29kb | +0.00b | 18.01s | -1.73m 🚀 |
| dist/heading1.ef0290b4.js | 8.26kb | +0.00b | 21.49s | +1.43s ⚠️ |
| dist/panel-error.8bc896be.js | 8.19kb | +0.00b | 18.01s | -1.72m 🚀 |
| dist/panel.6acb9fb3.js | 8.18kb | +0.00b | 19.77s | +1.32s ⚠️ |
| dist/table.500adb35.js | 8.18kb | +0.00b | 19.77s | +1.32s ⚠️ |
| dist/panel-success.7636d22e.js | 8.13kb | +0.00b | 18.01s | -1.73m 🚀 |
| dist/panel-note.2eea2289.js | 8.13kb | +0.00b | 18.01s | +1.23s ⚠️ |
| dist/16.073347bf.js | 5.24kb | +0.00b | 13.10s | +838.00ms ⚠️ |
| dist/16.c4ee6c00.js | 5.17kb | +0.00b | 12.01s | +899.00ms ⚠️ |
| dist/16.7f86346a.js | 4.89kb | +0.00b | 13.36s | +1.09s ⚠️ |
| dist/16.5f13de00.js | 4.84kb | +0.00b | 12.01s | +898.00ms ⚠️ |
| dist/16.51c4a0ba.js | 4.82kb | +0.00b | 14.55s | +1.41s ⚠️ |
| dist/16.c81c0695.js | 4.82kb | +0.00b | 14.55s | +1.41s ⚠️ |
| dist/16.7b85ceeb.js | 4.73kb | +0.00b | 13.10s | -2.19s 🚀 |
| dist/16.ba932c2e.js | 4.70kb | +0.00b | 13.10s | -2.19s 🚀 |
| dist/16.232fb9da.js | 4.69kb | +0.00b | 14.55s | +1.25s ⚠️ |
| dist/16.60895c0e.js | 4.65kb | +0.00b | 13.76s | +1.49s ⚠️ |
| dist/16.771934f3.js | 4.65kb | +0.00b | 12.01s | +899.00ms ⚠️ |
| dist/16.8a21dead.js | 4.64kb | +0.00b | 13.36s | +1.10s ⚠️ |
| dist/16.ba7ed30e.js | 4.64kb | +0.00b | 13.44s | +1.17s ⚠️ |
| dist/16.ea6e7d4b.js | 4.63kb | +0.00b | 12.01s | +900.00ms ⚠️ |
| dist/ak-editor.b0c28b13.css | 3.38kb | +0.00b | 1.98m | +21.60s ⚠️ |
| dist/EmojiUploadComponent.b1f515dc.js | 2.53kb | +0.00b | 35.59s | -2.46s 🚀 |
| dist/ResourcedEmojiComponent.4495d16e.js | 2.32kb | +0.00b | 10.82s | -1.39s 🚀 |
| dist/EmojiPickerComponent.39e7a3dc.js | 467.00b | +0.00b | 35.59s | -2.46s 🚀 |
| dist/index.html | 212.00b | +0.00b | 1.96m | +1.86m ⚠️ |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/262845681127916041)**
